[FSDP] Validate buffer dtype in pre-forward hook for FSDP mixed precision tests

vnalla-meta opened the pull request:
Summary: https://github.com/pytorch/pytorch/issues/104740

Test Plan: buck2 test mode/dev-nosan caffe2/test/distributed/fsdp:fsdp_mixed_precision -- test_full_precision_in_eval_buffers

Reviewed By: awgu

Differential Revision: D47858769

awgu reviewed (approved):
Approved internally

awgu commented:
@pytorchbot merge

awgu commented on test/distributed/fsdp/test_fsdp_mixed_precision.py:
```diff
@@ -889,16 +889,37 @@ def test_full_precision_in_eval_buffers(self):
+                    self.assertEqual(expected_dtype, buf.dtype)
+
+            def _get_underlying_module(m):
+                 return m.module if isinstance(m, FSDP) else m
```
```suggestion
                return m.module if isinstance(m, FSDP) else m
```

awgu commented on test/distributed/fsdp/test_fsdp_mixed_precision.py:
```diff
@@ -889,16 +889,37 @@ def test_full_precision_in_eval_buffers(self):
+                expected_dtype = (
+                    _BUFFER_ORIG_DTYPE if use_full_prec_in_eval else torch.float16
+                )
+                print([b for b in module.named_buffers()])
```
```suggestion
```

awgu commented on test/distributed/fsdp/test_fsdp_mixed_precision.py:
```diff
@@ -889,16 +889,37 @@ def test_full_precision_in_eval_buffers(self):
+                .register_forward_pre_hook(verify_eval_buffer_dtype)
+            )
+            hook_handles.append(
+                _get_underlying_module(fsdp_model[1])
+                .register_forward_pre_hook(verify_eval_buffer_dtype)
```
For lint:
```suggestion
                _get_underlying_module(fsdp_model[1]).register_forward_pre_hook(
                    verify_eval_buffer_dtype
                )
```

awgu commented on test/distributed/fsdp/test_fsdp_mixed_precision.py:
```diff
@@ -889,16 +889,37 @@ def test_full_precision_in_eval_buffers(self):
+
+            hook_handles = []
+            hook_handles.append(
+                _get_underlying_module(fsdp_model[0])
+                .register_forward_pre_hook(verify_eval_buffer_dtype)
```
For lint:
```suggestion
                _get_underlying_module(fsdp_model[0]).register_forward_pre_hook(
                    verify_eval_buffer_dtype
                )
```

vnalla-meta committed:
[FSDP] Validate buffer dtype in pre-forward hook for FSDP mixed precision tests (#106231)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/106231

https://github.com/pytorch/pytorch/issues/104740

Test Plan: buck2 test mode/dev-nosan caffe2/test/distributed/fsdp:fsdp_mixed_precision -- test_full_precision_in_eval_buffers

Reviewed By: awgu

Differential Revision: D47858769

fbshipit-source-id: 9d566155435f33f9e9d6152f3a3497ac8bd98aae

awgu commented:
@pytorchbot merge
